File: CSharp/One/Ast/AstHelper.cs
using One.Ast;
using System.Collections.Generic;

namespace One.Ast
{
    public class AstHelper {
        public static IInterface[] collectAllBaseInterfaces(IInterface intf)
        {
            var result = new Set<IInterface>();
            var toBeProcessed = new List<IInterface> { intf };
            
            while (toBeProcessed.length() > 0) {
                var curr = toBeProcessed.pop();
                result.add(curr);
                
                if (curr is Class class_ && class_.baseClass != null)
                    toBeProcessed.push((((ClassType)class_.baseClass)).decl);
                
                foreach (var baseIntf in curr.baseInterfaces)
                    toBeProcessed.push((((InterfaceType)baseIntf)).decl);
            }
            
            return Array.from(result.values());
        }
    }
}

File: PHP/One/Ast/AstHelper.php
<?php

namespace One\Ast\AstHelper;

use One\Ast\Types\IInterface;
use One\Ast\Types\Class_;
use One\Ast\AstTypes\InterfaceType;
use One\Ast\AstTypes\ClassType;

class AstHelper {
    static function collectAllBaseInterfaces($intf) {
        $result = new \OneLang\Set();
        $toBeProcessed = array($intf);
        
        while (count($toBeProcessed) > 0) {
            $curr = array_pop($toBeProcessed);
            $result->add($curr);
            
            if ($curr instanceof Class_ && $curr->baseClass !== null)
                $toBeProcessed[] = ($curr->baseClass)->decl;
            
            foreach ($curr->baseInterfaces as $baseIntf)
                $toBeProcessed[] = ($baseIntf)->decl;
        }
        
        return \OneLang\Array_::from($result->values());
    }
}
